Add http-add-on docs (#882)
kyma-project/keda-manager: docs/user/07-10-http-add-on.md

| **Operator** | Watches `HTTPScaledObject` resources and configures the Interceptor routing and KEDA ScaledObjects. |

**Request flow:**

![KEDA scaling scenario](../assets/scale-to-0.png "Scenario")

1. An `HTTP request` arrives at the **API Gateway**, which routes traffic to the **Interceptor's Service**.
2. The **Interceptor** receives the request, counts it, and queues it if the target workload has 0 replicas.
3. The **External Scaler** pulls request count metrics from the **Interceptor** via gRPC.
4. **KEDA** reads metrics from the **External Scaler** and scales the target **Deployment** accordingly (including to/from zero).
5. **KEDA** reconciles the **ScaledObject** (auto-created from HTTPScaledObject) to manage the scaling behavior.
6. The **KEDA-HTTP Operator** watches **HTTPScaledObject** resources and configures all add-on components (Interceptor routing, ScaledObject, External Scaler).
7. Once the **Deployment** has ready replicas, the **Interceptor** forwards the queued request to the application **Service**, which routes it to the running **Pod**.

## Enabling and Disabling the HTTP Add-on

> You enable the HTTP Add-on by annotating the Keda custom resource (CR):

Reviewer: ```suggestion
Enable the HTTP Add-on by annotating the Keda custom resource (CR):
```

| `keda.kyma-project.io/addon-namespace` | No | Namespace where the add-on is installed. Defaults to `kyma-system`. |
| `keda.kyma-project.io/addon-istio-injection` | No | Set to `true` to enable Istio sidecar injection on the add-on Deployments. Defaults to `false` — the add-on Deployments are annotated with `sidecar.istio.io/inject: "false"` unless this annotation is explicitly set to `true`. When enabled, the Interceptor Deployment also receives `traffic.sidecar.istio.io/excludeInboundPorts: "9090"` to prevent the sidecar from intercepting internal gRPC traffic. |

### Changing the Installation Namespace

To move the HTTP Add-on to a different namespace, update the `addon-namespace` annotation:

```bash
kubectl annotate keda default \
  keda.kyma-project.io/addon-namespace=my-new-namespace --overwrite
```

The controller detects the namespace change, removes only the HTTP Add-on resources from the old namespace (other Deployments, Services, etc. in that namespace are not affected), creates the new namespace, if it doesn't exist, with `istio-injection=enabled`, and installs the HTTP Add-on in the new namespace.

### Disabling the HTTP Add-on

> 

Reviewer: ```suggestion
To disable the HTTP add-on, run:
```

Reviewer: Make spaces between the title and the code itself.

### Interceptor Timeouts

The most important configuration options are the Interceptor's timeout settings. These control how long the Interceptor waits during a cold start and when to forward requests to your application.

| Environment Variable | Default | Description |
|---|---|---|
| `KEDA_HTTP_REQUEST_TIMEOUT` | `0s` (unlimited) | Total wall-clock deadline from request arrival to response completion. When `0`, there is no total request deadline — the request can wait indefinitely for scale-up. |
| `KEDA_HTTP_READINESS_TIMEOUT` | `0s` (unlimited) | How long to wait for the backing workload to have ≥1 ready replicas before giving up. When `0`, the readiness wait is bounded only by the request timeout, giving the full request budget to cold starts. |
| `KEDA_HTTP_RESPONSE_HEADER_TIMEOUT` | `300s` | How long to wait for response headers from the backend after the request is forwarded. Acts as a safety net against hung backends. Set to `0` to disable. |
| `KEDA_HTTP_CONNECT_TIMEOUT` | `500ms` | Per-attempt TCP dial timeout when connecting to the backend. Bounded by the request context deadline. |

> > [!Note]

Reviewer: ```suggestion
> [!NOTE]
```

> > [!Note]

Reviewer: ```suggestion
> [!NOTE]
```

# KEDA HTTP Add-on

## Overview

The [KEDA HTTP Add-on](https://github.com/kedacore/http-add-on) extends KEDA with the ability to scale HTTP workloads to and from zero based on incoming request rate. It works by placing an **Interceptor proxy** in front of your application that counts, queues, and forwards HTTP requests — enabling true scale-to-zero without losing any requests.

## Architecture

The HTTP Add-on consists of three components:

| Component | Role |
|---|---|
| **Interceptor** | A reverse proxy that sits in front of your application. It counts incoming requests, queues them when the target has 0 replicas, and forwards them once the target is ready. |
| **Scaler** (External Scaler) | Exposes request-rate metrics to KEDA via gRPC so KEDA can make scaling decisions. |
| **Operator** | Watches `HTTPScaledObject` resources and configures the Interceptor routing and KEDA ScaledObjects. |

**Request flow:**

> ![KEDA scaling scenario](../assets/scale-to-0.png "Scenario")

Reviewer: The diagram mentions Keda twice. Are they both the same? Are we talking about [KEDA](https://keda.sh/) the autoscaler itself, or the Keda module?

<img width="756" height="165" alt="Image" src="https://github.com/user-attachments/assets/bf2e0ff4-88d4-4f5d-bf26-c945e2e5451b" />

Author: The green one is a visual grouping box that represents the KEDA autoscaler component group, similar to how the other green area is labeled "HTTP add-on".
Keda box inside is actual  Keda the autocaler itself - I suppose we should rename it to make it more clear, could you suggest any name, e.g. KEDA?

Author: We've changed green Keda to `KEDA autocaler component group`, and Keda in box to `KEDA`

1. An `HTTP request` arrives at the **API Gateway**, which routes traffic to the **Interceptor's Service**.
2. The **Interceptor** receives the request, counts it, and queues it if the target workload has 0 replicas.
3. The **External Scaler** pulls request count metrics from the **Interceptor** via gRPC.
4. **KEDA** reads metrics from the **External Scaler** and scales the target **Deployment** accordingly (including to/from zero).
5. **KEDA** reconciles the **ScaledObject** (auto-created from HTTPScaledObject) to manage the scaling behavior.
6. The **KEDA-HTTP Operator** watches **HTTPScaledObject** resources and configures all add-on components (Interceptor routing, ScaledObject, External Scaler).
7. Once the **Deployment** has ready replicas, the **Interceptor** forwards the queued request to the application **Service**, which routes it to the running **Pod**.

> ## Enabling the HTTP Add-on

Reviewer: ```suggestion
## Enabling, Upgrading and Disabling the HTTP Add-on
```

> You enable the HTTP Add-on by annotating the Keda CR:

Reviewer: ```suggestion
You enable the HTTP Add-on by annotating the Keda custom resource (CR):
```

  keda.kyma-project.io/addon-namespace=keda
```

### Annotations Reference

| Annotation | Required | Description |
|---|---|---|
| `keda.kyma-project.io/addon-enabled` | Yes | Set to `true` to install, `false` to uninstall. |
| `keda.kyma-project.io/addon-version` | No | Semver version of the HTTP Add-on to install (for example, `0.13.0`). If omitted, the latest version is resolved automatically. |
| `keda.kyma-project.io/addon-namespace` | No | Namespace where the add-on is installed. Defaults to `kyma-system`. |

### Upgrading the HTTP Add-on Version

To upgrade (or downgrade) the HTTP Add-on to a different version, update the `addon-version` annotation:

```bash
kubectl annotate keda default \
  keda.kyma-project.io/addon-version=0.14.0 --overwrite
```

The controller automatically detects the version change, removes the old version's resources, and installs the new version in the same namespace. No manual cleanup is required.

### Changing the Installation Namespace

To move the HTTP Add-on to a different namespace, update the `addon-namespace` annotation:

```bash
kubectl annotate keda default \
  keda.kyma-project.io/addon-namespace=my-new-namespace --overwrite
```

> The controller detects the namespace change and:

Reviewer: ```suggestion
The controller detects the namespace change, removes only the HTTP Add-on resources from the old namespace (other Deployments, Services, etc. in that namespace are not affected), creates the new namespace, if it doesn't exist, with `istio-injection=enabled`, and installs the HTTP Add-on in the new namespace.
```

> 1. Removes **only the HTTP Add-on resources** from the old namespace (other Deployments, Services, etc. in that namespace are not affected).

Reviewer: ```suggestion
```

> 2. Creates the new namespace (if it doesn't exist) with `istio-injection=enabled`.

Reviewer: ```suggestion
```

> 3. Installs the HTTP Add-on in the new namespace.

Reviewer: ```suggestion
```

### Disabling the HTTP Add-on

```bash
kubectl annotate keda default \
  keda.kyma-project.io/addon-enabled=false --overwrite
```

> This removes all add-on resources from the cluster (only the resources managed by the HTTP Add-on — other workloads in the namespace are not affected).

Reviewer: ```suggestion
This removes all add-on resources from the cluster. Only the resources managed by the HTTP Add-on are removed. Other workloads in the namespace are not affected.
```

## Configuring the HTTP Add-on

> The HTTP Add-on components are configured via environment variables on their Deployments. You can customize them by patching the respective Deployment after installation.

Reviewer: ```suggestion
The HTTP Add-on components are configured using environment variables on their Deployments. You can customize them by patching the respective Deployment after installation.
```

### Interceptor Timeouts

> The most important configuration options are the Interceptor's timeout settings. These control how long the Interceptor waits during cold start and when forwarding requests to your application.

Reviewer: ```suggestion
The most important configuration options are the Interceptor's timeout settings. These control how long the Interceptor waits during a cold start and when to forward requests to your application.
```

| Environment Variable | Default | Description |
|---|---|---|
| `KEDA_HTTP_REQUEST_TIMEOUT` | `0s` (unlimited) | Total wall-clock deadline from request arrival to response completion. When `0`, there is no total request deadline — the request can wait indefinitely for scale-up. |
| `KEDA_HTTP_READINESS_TIMEOUT` | `0s` (unlimited) | How long to wait for the backing workload to have ≥1 ready replicas before giving up. When `0`, the readiness wait is bounded only by the request timeout, giving the full request budget to cold starts. |
| `KEDA_HTTP_RESPONSE_HEADER_TIMEOUT` | `300s` | How long to wait for response headers from the backend after the request is forwarded. Acts as a safety net against hung backends. Set to `0` to disable. |
| `KEDA_HTTP_CONNECT_TIMEOUT` | `500ms` | Per-attempt TCP dial timeout when connecting to the backend. Bounded by the request context deadline. |

> > **Note:** If `KEDA_HTTP_REQUEST_TIMEOUT` is set to `0` (default), the Interceptor will wait **indefinitely** for the target to scale up. This is the recommended setting when using the EnvoyFilter retry policy, as the retry policy on the Ingress Gateway side handles client-facing timeouts.

Reviewer: ```suggestion
> [!NOTE] 
> If `KEDA_HTTP_REQUEST_TIMEOUT` is set to `0` (default), the Interceptor waits indefinitely for the target to scale up. This is the recommended setting when using the EnvoyFilter retry policy, as the retry policy on the Ingress Gateway side handles client-facing timeouts.
```

### Interceptor Connection Pool

> These settings control the Interceptor's internal HTTP connection pool to backend services:

Reviewer: ```suggestion
The following settings control the Interceptor's internal HTTP connection pool to backend services:
```

| Environment Variable | Default | Description |
|---|---|---|
| `KEDA_HTTP_MAX_IDLE_CONNS` | `1000` | Max idle connections across all backend services. Increase if you proxy to many unique backends. |
| `KEDA_HTTP_MAX_IDLE_CONNS_PER_HOST` | `200` | Max idle connections per backend service. Increase if you observe many new connection establishments under load. |
| `KEDA_HTTP_FORCE_HTTP2` | `false` | Force HTTP/2 for all upstream connections. |

### Interceptor Behavior

| Environment Variable | Default | Description |
|---|---|---|
| `KEDA_HTTP_ENABLE_COLD_START_HEADER` | `true` | When enabled, the Interceptor adds an `X-KEDA-HTTP-Cold-Start: true` response header if the request triggered a scale-from-zero. Useful for observability. |
| `KEDA_HTTP_LOG_REQUESTS` | `false` | Log every incoming request (for debugging). |

### Scaler Configuration

> The External Scaler component has these key settings:

Reviewer: ```suggestion
The External Scaler component has the following key settings:
```

| Environment Variable | Default | Description |
|---|---|---|
| `KEDA_HTTP_QUEUE_TICK_DURATION` | `500ms` | How often the scaler fetches request counts from the Interceptor. Lower values give faster scaling reactions but increase gRPC traffic. |
| `KEDA_HTTP_SCALER_STREAM_INTERVAL_MS` | `200` | Interval (ms) between metric stream updates sent to KEDA. |

## Usage: HTTPScaledObject

After the add-on is installed, create an `HTTPScaledObject` to configure scaling for your workload:

```yaml
apiVersion: http.keda.sh/v1alpha1
kind: HTTPScaledObject
metadata:
  name: my-app
  namespace: my-namespace
spec:
  hosts:
  - "my-app.example.com"
  pathPrefixes:
  - /
  scaleTargetRef:
    name: my-app
    kind: Deployment
    apiVersion: apps/v1
    service: my-app
    port: 8080
  replicas:
    min: 0
    max: 10
  scalingMetric:
    requestRate:
      targetValue: 10
      granularity: "1s"
      window: "1m"
```

Key fields:
- **`replicas.min: 0`** — enables scale-to-zero.
- **`scalingMetric.requestRate.targetValue`** — number of requests per second per replica that triggers scale-out.
- **`scalingMetric.requestRate.window`** — time window over which request rate is averaged.

> ## Cold Start and the EnvoyFilter Requirement

Reviewer: This whole section looks like a troubleshooting guide to me. Can you create a `troubleshooting-guide` folder in the `user` folder and put it there? See [`serverless`](https://github.com/kyma-project/serverless/tree/main/docs/user/troubleshooting-guides) for reference and template (Symptom-Cause-Solution).

Author: Thanks, we will move it to sth like you mentioned
